workflow/scripts/add_district_heating_subnodes.py

```diff
@@ -0,0 +1,421 @@
+        n.loads.loc[
+            f'{subnode["cluster"]} low-temperature heat for industry', "p_set"
+        ] *= (1 - demand_ratio * lti_share)
+
+        # Replicate district heating stores and links of mother node for subnodes
+        n.madd(
+            "Bus",
+            [f"{subnode['cluster']} {subnode['Stadt']} urban central water tanks"],
+            location=f"{subnode['cluster']} {subnode['Stadt']}",
+            carrier="urban central water tanks",
+            unit="MWh_th",
+        )
+
+        stores = (
+            n.stores.filter(like=f"{subnode['cluster']} urban central", axis=0)
+            .reset_index()
+            .replace(
+                {
+                    f"{subnode['cluster']} urban central": f"{subnode['cluster']} {subnode['Stadt']} urban central"
+                },
+                regex=True,
+            )
+            .set_index("Store")
+        )
+        n.madd("Store", stores.index, **stores)
+
+        links = (
+            n.links.loc[~n.links.carrier.str.contains("heat pump")]
+            .filter(like=f"{subnode['cluster']} urban central", axis=0)
+            .reset_index()
+            .replace(
+                {
+                    f"{subnode['cluster']} urban central": f"{subnode['cluster']} {subnode['Stadt']} urban central"
+                },
+                regex=True,
+            )
+            .set_index("Link")
+        )
+        n.madd("Link", links.index, **links)
+
+        # Add heat pumps to subnode
+        heat_pumps = (
+            n.links.filter(
+                regex=f"{subnode['cluster']} urban central.*heat pump", axis=0
+            )
+            .reset_index()
+            .replace(
+                {
+                    f"{subnode['cluster']} urban central": f"{subnode['cluster']} {subnode['Stadt']} urban central"
+                },
+                regex=True,
+            )
+            .set_index("Link")
+        ).drop("efficiency", axis=1)
+        heat_pumps_t = n.links_t.efficiency.filter(
+            regex=f"{subnode['cluster']} urban central.*heat pump"
+        )
+        heat_pumps_t.columns = heat_pumps_t.columns.str.replace(
+            f"{subnode['cluster']} urban central",
+            f"{subnode['cluster']} {subnode['Stadt']} urban central",
+        )
+        n.madd("Link", heat_pumps.index, efficiency=heat_pumps_t, **heat_pumps)
+
+        # Add heat vent to subnode
+        n.madd(
+            "Generator",
+            [f"{name} heat vent"],
+            bus=name,
+            location=f"{subnode['cluster']} {subnode['Stadt']}",
+            carrier="urban central heat vent",
+            p_nom_extendable=True,
+            p_min_pu=-1,
+            p_max_pu=0,
+            unit="MWh_th",
+        )
+
+    return
+
+
+def extend_cops(cops: xr.DataArray, subnodes: gpd.GeoDataFrame) -> xr.DataArray:
+    """
+    Extend COPs (Coefficient of Performance) by subnodes mirroring the timeseries of the corresponding
```

Comment:
Could we write the docstring in numpy style?

Reply:
Done

Reply:
Looks like you've missed that one:)

Reply:
Done